src/mscorlib/src/System/Reflection/Emit/CustomAttributeBuilder.cs

```diff
@@ -176,114 +186,137 @@ private bool ValidateType(Type t)
             for (i = 0; i < namedProperties.Length; i++)
             {
                 // Validate the property.
-                if (namedProperties[i] == null)
+                PropertyInfo property = namedProperties[i];
+                if (property == null)
                     throw new ArgumentNullException("namedProperties[" + i + "]");
 
                 // Allow null for non-primitive types only.
-                Type propType = namedProperties[i].PropertyType;
-                if (propertyValues[i] == null && propType.IsPrimitive)
+                Type propType = property.PropertyType;
+                object propertyValue = propertyValues[i];
+                if (propertyValue == null && propType.IsValueType)
                     throw new ArgumentNullException("propertyValues[" + i + "]");
 
                 // Validate property type.
                 if (!ValidateType(propType))
                     throw new ArgumentException(Environment.GetResourceString("Argument_BadTypeInCustomAttribute"));
 
                 // Property has to be writable.
-                if (!namedProperties[i].CanWrite)                
+                if (!property.CanWrite)
                     throw new ArgumentException(Environment.GetResourceString("Argument_NotAWritableProperty"));
-                    
+
+                if (!(property is PropertyBuilder) && property.GetIndexParameters().Length != 0)
```

Comment:
This is a breaking change. I agree with fixing NullReferenceExceptions and Debug.Asserts that occur within the CustomAttributeBuilder constructor call, but in this case (and in the case of the non-writable fields and properties), this change is introducing an exception into a call that previously succeded. (Even though in the end, it ends up writing an illegal custom attribute - still, there may be existing code that relies on that - they write out an attribute and never get around to reading it back as an instantiated Attribute, or they may be being used to generate test data.)

I'd recommend scoping this PR down to the fixes that fall within the list of <a href="https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md">allowed changes</a>. If there a strong case for fixing the rest, please indicate their breaking nature in the issue and mark those issues "api-ready-for-review".